File: Tree/MinHeap/MinHeap.py
class MinHeap:
    def __init__(self):
        self.heap = []

    def size(self):
        return len(self.heap)

    def peek(self):
        if self.size() == 0:
            return None
        return self.heap[0]

    def _parent(self, i): return (i - 1) // 2
    def _left_child(self, i): return (2 * i) + 1
    def _right_child(self, i): return (2 * i) + 2

    def insert(self, value):
        self.heap.append(value)
        self._heapify_up(self.size() - 1)

    def _heapify_up(self, i):
        while i > 0 and self.heap[i] < self.heap[self._parent(i)]:
            parent_idx = self._parent(i)
            # Swap
            self.heap[i], self.heap[parent_idx] = self.heap[parent_idx], self.heap[i]
            i = parent_idx

    def extract_min(self):
        if self.size() == 0:
            return None
        if self.size() == 1:
            return self.heap.pop()

        root_val = self.heap[0]
        # Move the last element to the root position
        self.heap[0] = self.heap.pop()
        self._heapify_down(0)
        return root_val

    def _heapify_down(self, i):
        smallest = i
        left = self._left_child(i)
        right = self._right_child(i)

        if left < self.size() and self.heap[left] < self.heap[smallest]:
            smallest = left
        if right < self.size() and self.heap[right] < self.heap[smallest]:
            smallest = right

        if smallest != i:
            self.heap[i], self.heap[smallest] = self.heap[smallest], self.heap[i]
            self._heapify_down(smallest)

    def print_heap(self):
        print("Heap Array Layout:", self.heap)

if __name__ == "__main__":
    heap = MinHeap()
    elements = [15, 10, 20, 5, 8]
    for x in elements:
        heap.insert(x)

    heap.print_heap()  # Expected root: 5
    print("Minimum element picked (peek):", heap.peek())
    
    print("Extracted min:", heap.extract_min())  # Removes 5
    heap.print_heap()  # New root should be 8

File: Tree/MinHeap/MinHeap.cpp
#include <iostream>
#include <vector>
#include <stdexcept>

class MinHeap {
private:
    std::vector<int> heap;

    int parent(int i) const { return (i - 1) / 2; }
    int leftChild(int i) const { return (2 * i) + 1; }
    int rightChild(int i) const { return (2 * i) + 2; }

    void heapifyUp(int i) {
        while (i > 0 && heap[i] < heap[parent(i)]) {
            std::swap(heap[i], heap[parent(i)]);
            i = parent(i);
        }
    }

    void heapifyDown(int i) {
        int smallest = i;
        int left = leftChild(i);
        int right = rightChild(i);

        if (left < size() && heap[left] < heap[smallest]) smallest = left;
        if (right < size() && heap[right] < heap[smallest]) smallest = right;

        if (smallest != i) {
            std::swap(heap[i], heap[smallest]);
            heapifyDown(smallest);
        }
    }

public:
    int size() const { return heap.size(); }
    bool isEmpty() const { return heap.empty(); }

    void insert(int value) {
        heap.push_back(value);
        heapifyUp(size() - 1);
    }

    int extractMin() {
        if (isEmpty()) throw std::underflow_error("Heap is empty");
        
        int rootVal = heap[0];
        heap[0] = heap.back();
        heap.pop_back();
        
        if (!isEmpty()) heapifyDown(0);
        return rootVal;
    }

    int peek() const {
        if (isEmpty()) throw std::underflow_error("Heap is empty");
        return heap[0];
    }

    void print() const {
        std::cout << "Heap Array representation: ";
        for (int x : heap) std::cout << x << " ";
        std::cout << "\n";
    }
};

int main() {
    MinHeap mh;
    mh.insert(15);
    mh.insert(10);
    mh.insert(20);
    mh.insert(5);
    
    mh.print();
    std::cout << "Extracted min: " << mh.extractMin() << "\n";
    mh.print();

    return 0;
}
